javalib/src/main/scala/java/net/URLEncoder.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.net
+
+import scala.annotation.switch
+
+import java.io.UnsupportedEncodingException
+import java.nio.{CharBuffer, ByteBuffer}
+import java.nio.charset.{Charset, CharsetDecoder}
+
+import java.util.ScalaOps._
+import java.nio.charset.CodingErrorAction
+
+object URLEncoder {
+  final val EncodeAsIsLength = 128
+
+  private val EncodedAsIs: Array[Boolean] = {
+    val r = new Array[Boolean](EncodeAsIsLength) // initialized with false
+    r('.') = true
+    r('-') = true
+    r('*') = true
+    r('_') = true
+    for (c <- '0'.toInt to '9'.toInt)
+      r(c) = true
+    for (c <- 'A'.toInt to 'Z'.toInt)
+      r(c) = true
+    for (c <- 'a'.toInt to 'z'.toInt)
+      r(c) = true
+    r
+  }
+
+  private val PercentEncoded: Array[String] = {
+    val hexDigits = "0123456789ABCDEF"
+    val r = new Array[String](256)
+    for (b <- 0 until 256)
+      r(b) = "%" + hexDigits.charAt(b >>> 4) + hexDigits.charAt(b & 0xf)
+    r
+  }
+
+  @Deprecated
+  def encode(s: String): String = encode(s, Charset.defaultCharset())
+
+  def encode(s: String, enc: String): String = {
+    if (!Charset.isSupported(enc))
+      throw new UnsupportedEncodingException(enc)
+    encode(s, Charset.forName(enc))
+  }
+
+  def encode(s: String, charset: Charset): String = {
+    val EncodedAsIs = this.EncodedAsIs // local copy
+
+    @inline def encodeAsIs(c: Char): Boolean =
+      c < EncodeAsIsLength && EncodedAsIs(c)
+
+    @inline def encodeUsingCharset(c: Char): Boolean =
+      c != ' ' && !encodeAsIs(c)
+
+    var len = s.length()
+    var i = 0
+
+    while (i != len && encodeAsIs(s.charAt(i)))
+      i += 1
+
+    if (i == len) {
+      s
+    } else {
+      val PercentEncoded = this.PercentEncoded // local copy
+
+      val charBuffer = CharBuffer.wrap(s)
+      val encoder = charset.newEncoder().onUnmappableCharacter(CodingErrorAction.REPLACE)
+      val bufferArray = new Array[Byte](((len - i + 1) * encoder.maxBytesPerChar()).toInt)
+      val buffer = ByteBuffer.wrap(bufferArray)
```

Comment:
Use ByteBuffer.allocate?

Reply:
The idea was to guarantee that there is an `Array` to perform direct access on it, rather than adding in the cost of the `ByteBuffer` safety for each `get()`. This is not unusual: algorithms using `Buffer` even have dedicated codepaths when they can access the underlying array.